test_data/std/bus_permutation_via_challenges.asm

```diff
@@ -0,0 +1,32 @@
+use std::prover::Query;
+use std::convert::fe;
+use std::protocols::permutation_via_bus::permutation;
+use std::math::fp2::from_base;
+use std::prover::challenge;
+
+machine Main with degree: 8 {
+    let alpha = from_base(challenge(0, 1));
+    let beta = from_base(challenge(0, 2));
+
+    col fixed first_four = [1, 1, 1, 1, 0, 0, 0, 0];
+
+    // Two pairs of witness columns, claimed to be permutations of one another
+    // (when selected by first_four and (1 - first_four), respectively)
+    col witness a1(i) query Query::Hint(fe(i));
+    col witness a2(i) query Query::Hint(fe(i + 42));
+    col witness b1(i) query Query::Hint(fe(7 - i));
+    col witness b2(i) query Query::Hint(fe(7 - i + 42));
+
+    let permutation_constraint = Constr::Permutation(
+        (Option::Some(first_four), Option::Some(1 - first_four)),
+        [(a1, b1), (a2, b2)]
+    );
+
+    // TODO: Functions currently cannot add witness columns at later stages,
+    // so we have to manually create it here and pass it to permutation(). 
+    col witness stage(1) z;
+    col witness stage(1) u;
```

Comment:
```suggestion
    col witness stage(1) u;
    
    // For this test, instead of exposing the final accumulators as public and adding a check to the verifier,
    // we add a constraint that checks that they are equal.
    is_first' * (z - u) = 0;
```

What do you think of this? The benefits I see is that:
- We already test this now (have you manually checked that it does produce the same accumulators?)
- Even once the accumulators are exposed as publics, I think this fits better into our testing framework. For example, the `test_halo2` function does not let us easily add additional verifier checks.

Reply:
This is far better to constrain the equation than to expose public values, and then test it.

- Yes, I have manually checked that their sum is equal to zero (they don't need to equal each other, do they?)
- By constraining it, we can be sure that the permutation implemented via bus is correct without having to look for the verifier.

However, I don't think the correct form of constraint is `is_first' * (z - u) = 0`. Instead, we need to check that their sum is equal to zero. So, the correct form should be `is_first' * (z + u) = 0` in my opinion. What do you think?

Reply:
Yes, right!